### PR TITLE
Remove newScheduler for reducing complexity

### LINKED ISSUE #108788: unroll scheduler.New and remove Configurator
### What would you like to be added?

1. `Configurator` seems redundant today as it only holds a method of `create` initializing several components, so I'd like to unroll `scheduler.New`  by moving the `create` logic to `scheduler.New`, then `Configurator` can be removed. The benefits: 
a) No need to maintain the `Configurator` struct
b) Take all the logics in a glance, easy to read. Before, `create` did a lot of things. https://github.com/kubernetes/kubernetes/blob/aface3ab99b5c8c6ac8c4892c8019742075a461d/pkg/scheduler/factory.go#L49-L86

3. What's more, `newScheduler` function is hard to maintain especially in tests, I'd like to refactor this with closure like `func WithProfiles(p ...schedulerapi.KubeSchedulerProfile) Option` we used in `New` scheduler. https://github.com/kubernetes/kubernetes/blob/aface3ab99b5c8c6ac8c4892c8019742075a461d/pkg/scheduler/scheduler_test.go#L2435-L2445

So I'd like to separate this into two subtasks:

- [x] unroll `scheduler.New` and remove `Configurator`
- [ ] refactor `newScheduler` with closure

### Why is this needed?

Easy to maintain codes.

### PR DESCRIPTION
**Pull request by kerthcet**
Signed-off-by: kerthcet <kerthcet@gmail.com>

#### What type of PR is this?
/kind cleanup
/sig scheduling

#### What this PR does / why we need it:

For readability and maintainability.

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubernetes/issues/108788

#### Does this PR introduce a user-facing change?

```release-note
None
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by kerthcet** [comment]
cc @kubernetes/sig-scheduling-leads

**Review by alculquicondor — COMMENTED** [review]
I think it's more common in the codebase that we use `withX`

**Comment by kerthcet** [comment]
> I think it's more common in the codebase that we use `withX`

Yes, that's what came into my mind first. But I found we already have `schedulerOptions` under the package. It seems conflicts in names and easy to confuse.

This pattern is widely used in cli. https://github.com/kubernetes/kubernetes/blob/9c9b29032f76e39b0e36200c88fe2f127389a9fe/staging/src/k8s.io/cli-runtime/pkg/resource/builder.go#L730-L733

**Comment by alculquicondor** [comment]
oh right... This seems to be another legacy leftover from the times when the scheduler was split in two objects.

We should just get rid of `newScheduler` instead (and migrate the tests that use it to `New`).

File: pkg/scheduler/scheduler.go [hunk 1]
```diff
@@ -98,6 +98,52 @@ type Scheduler struct {
...
+func (s *Scheduler) profiles(profiles profile.Map) *Scheduler {
+	s.Profiles = profiles
+	return s
+}
+
+func (s *Scheduler) do() *Scheduler {
+	s.SchedulePod = s.schedulePod
+	s.FailureHandler = s.handleSchedulingFailure
+	return s
+}
```
Comment (Huang-Wei) [inline pkg/scheduler/scheduler.go:145]: `do()` doesn't read well IMO. Maybe `apply` or `defaults`.

**Comment by kerthcet** [comment]
> oh right... This seems to be another legacy leftover from the times when the scheduler was split in two objects.
> 
> We should just get rid of `newScheduler` instead (and migrate the tests that use it to `New`).

It would be a little difficult, `scheduler.New()` helps to instantiate a scheduler with API-level options, like apiVersion, KubeSchedulerProfile, but the `newScheduler` here instantiates with internal parameters, they're from different layer. As a result, lots of unit tests depend on `newScheduler` to mock the server, I think it works well.

File: pkg/scheduler/scheduler.go [hunk 2]
```diff
@@ -98,6 +98,52 @@ type Scheduler struct {
 	nextStartNodeIndex int
 }
 
+// newScheduler creates a Scheduler object.
+func newScheduler(
```
[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/scheduler/scheduler.go:102]: Yes, these 4 parameters are widely used in summary.

File: pkg/scheduler/scheduler.go (hunk 1 above)
Comment (kerthcet) [reply pkg/scheduler/scheduler.go:145]: Change to apply()

File: pkg/scheduler/scheduler.go [hunk 3]
```diff
@@ -98,6 +98,52 @@ type Scheduler struct {
...
+	}
+	return &sched
+}
+
+func (s *Scheduler) extenders(extenders []framework.Extender) *Scheduler {
+	s.Extenders = extenders
+	return s
+}
+
+func (s *Scheduler) nextPod(fc func() *framework.QueuedPodInfo) *Scheduler {
```
[the remark this answers is not in the document]
Comment (kerthcet) [reply pkg/scheduler/scheduler.go:121]: done

**Comment by alculquicondor** [comment]
I'm against adding more setup complexity just because unit tests use the low level builder.

I think changing the unit tests is the right direction.

File: pkg/scheduler/scheduler.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (ahg-g) [reply pkg/scheduler/scheduler.go:102]: I suggest the following: remove the `newScheduler` function and create an `applyDefaultHandlers` function for `Scheduler` that sets `SchedulePod` and `FailureHandler`.

In `New` and the tests, just create a `Scheduler` struct and invoke `applyDefaultHandlers` right after (unless the handlers needs to be set to something different).

**Comment by kerthcet** [comment]
> I'm against adding more setup complexity just because unit tests use the low level builder.
> 
> I think changing the unit tests is the right direction.

In unit tests, scheduler should not care much about api level fields, e.g. in extender tests, we use the framework extender which is an interface, it's easy to mock the implementation https://github.com/kubernetes/kubernetes/blob/575031b68f5d52e541de6418a59a832252244486/pkg/scheduler/extender_test.go#L43-L51

But if we change to api lever extender, we have to define many fields, it's verbosely. https://github.com/kubernetes/kubernetes/blob/575031b68f5d52e541de6418a59a832252244486/pkg/scheduler/apis/config/types.go#L246-L252

**Comment by kerthcet** [comment]
To reduce the complexity, I'll follow @ahg-g advices, create a `Scheduler` directly.

**Comment by ahg-g** [comment]
please squash

**Comment by kerthcet** [comment]
Squashed @ahg-g

**Comment by ahg-g** [comment]
/approve
/hold
leaving lgtm to Wei or Aldo

**Review by alculquicondor — COMMENTED** [review]
/lgtm
/approve
way better :smiley:

**Pull request merged**